### DISCUSSION TITLE
Click on a row item in a table and forward to another page?

### DISCUSSION THREAD

**Discussion by JustinHowes01** [zauberzeug/nicegui#802] [Q&A] [answered]
### Question

I have a table that includes IP addresses, and I want to be able to click on the IP address, which will then forward the user to said IP.

1. How would I make the IP address in the table be a clickable link, and if that's not possible, what other alternatives could I use?

2. I have a function that redirects the user to the page with the IP address, as seen in the code below:
    ```py
    def individual_server_redirect(ip_address):
        ui.open(f'/servers/{ip_address}')
        return
    ```
    How would I extract the IP address from the link in order to enter it into the function? (The IP address is also the key of the rows.)

**Comment by falkoschindler** [comment]
Hi @JustinHowes01!
Getting row click events of a `ui.table` is currently a bit complicated (https://github.com/zauberzeug/nicegui/discussions/664).
Alternatively you can use a combination of `q-list` and `q-item` (see https://github.com/zauberzeug/nicegui/issues/698#issuecomment-1493688706) or simply `ui.row`s with arbitrary content:

```py
for ip in ['192.168.0.42', '192.168.0.43', '192.168.0.45']:
    with ui.row().classes('items-center'):
        ui.label(f'IP {ip}')
        ui.button(on_click=lambda ip=ip: ui.open(ip)).props('flat fab-mini icon=open_in_new')
```

But I don't quite understand point 2. Why do you need to "extract" the address? Maybe you can show a bit more of your code.

**Reply by JustinHowes01** [reply]
Sorry about that!

Here's some more code:
```py
with ui.table(columns=columns,
              rows=rows,
              row_key='ip',
              selection='multiple',
              pagination=50,
              on_select=lambda e: update_selection(e.selection)).classes('bg-[#E0E0E0] w-full') as table:
    with table.add_slot('top'):
        with ui.row().classes('items-center justify-between border-b-4 border-[#a1a1aa] pb-4 w-full'):
            with ui.row().classes('gap-2.5'):
                Buttons.control_button('Enable Mining', 'green', selection)
                Buttons.control_button('Disable Mining', 'red', selection)
                Buttons.control_button('Reboot', 'red', selection)
            with ui.row():
                with ui.input(placeholder='Search').props('type=search').bind_value(table, 'filter').add_slot('append'):
                    ui.icon('search')
```

And this is the result from the code:
![image](https://user-images.githubusercontent.com/69179857/233744584-d6a6a517-ac6b-4c8e-bf1e-4f3ef84a9b14.png)

I'm wanting to figure out how to redirect the user to a page containing information on a server by either clicking on the IP address itself, or clicking a button that redirects them using the function in my original comment. To elaborate on "extracting" the IP address, I just need the IP address from the item on the table in order to actually redirect the user, unless there is a better way to do so.

Also, the `Buttons.control_button()` item in the above code is just an object oriented button that takes the selection of the table and sends it to the backend along with the command.

Thank you very much for helping me btw!

**Reply by falkoschindler** [reply, marked as answer]
One way would be to define the HTML for the IP address cells. This can be done using the "body-cell-[name]" slot where "[name]" is your "ip" column. The slot can use the argument "props" which contains the current row. Adding an anchor tag `<a>` to the table cell `<td>` is probably the easiest way to make the IP address redirect to the server:
```py
table.add_slot('body-cell-ip', r'''
    <td :props="props">
        <a :href="'https://' + props.row.ip">{{props.row.ip}}</a>
    </td>
''')
```

**Reply by JustinHowes01** [reply]
That's perfect! Thank you very much.

I do have another question though. I see in Quasar's table docs where the `body-cell-[name]` object comes from.
![image](https://user-images.githubusercontent.com/69179857/233860276-f4eb26a0-de7c-468f-85a4-b2630e47593c.png)

However, I'm having an issue where I read the documentation, but even though I see the information that I need, I'm having trouble implementing it in NiceGUI (I would have never figured that out). This is not any issue with NiceGUI, because it's amazing, but more of an issue with my lack of experience with frontend development and a disconnect between Python syntax and knowledge of frontend code, as such in your example.

Are there any examples you may be able to recommend so I can see more NiceGUI syntax with more configuration using frontend code similar to your example?

**Reply by falkoschindler** [reply]
@JustinHowes01 I just learned about the "body-cell-[name]" slot yesterday too. Having Vue, Quasar, Tailwind, CSS, JavaScript and HTML right at our fingertips opens vast possibilities, but there're always new things to discover. The overall goal of NiceGUI is to integrate more and more features so that they become part of the official API and documentation, and become accessible for everyone without deeper knowledge about web development. But for some things like scoped slots this is really difficult. And our users keep pushing the limit of what is possible with NiceGUI, so there's quite some frontend knowledge involved for the more advanced examples.

To learn about scoped slots, the QTable documentation https://quasar.dev/vue-components/table in combination with NiceGUI's table documentation https://nicegui.io/documentation/table is a good starting point. There's also a slot example at https://github.com/zauberzeug/nicegui/blob/main/examples/slots/main.py. Last but not least, you're always welcome to ask questions here at GitHub or on Discord. This way we try to spread the knowledge, grow the community and ultimately increase the amount of googleable NiceGUI code out there. 😉
